src/v/utils/container_utils.h

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright 2022 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "seastarx.h"
+#include "utils/concepts-enabled.h"
+
+#include <seastar/core/future.hh>
+#include <seastar/coroutine/maybe_yield.hh>
+
+/// Clears the given container while yielding to seastar after every
+/// element is erased. This is used to avoid reactor stall warnings
+/// when deleting containers with a large number of objects.
+///
+/// \param c a container to be erased.
+template<typename Container>
+CONCEPT(requires requires(Container c, typename Container::iterator i) {
+    { c.begin() } -> std::same_as<typename Container::iterator>;
+    { c.end() } -> std::same_as<typename Container::iterator>;
+    { c.erase(i, ++i) } -> std::same_as<typename Container::iterator>;
+})
+ss::future<> async_clear(Container& c) {
+    auto current = c.begin();
+
+    while (current != c.end()) {
+        // Note: abseil containers not compatible with std containers
+        // in all other overloads of erase.
+        current = c.erase(current, ++current);
```

Comment:
Unfortunately abseil's flat_hash_map container doesn't have api support for easily chunking the data structure. For now I've added a size threshold to the function to call `clear` when the container is below a certain size.

Reply:
Correct me if i'm wrong but i think John was just mentioning maybe it would be a good idea to consider performing the yield after a few iterations, not on every one. Maybe the loop could calculate how much time has elapsed between clears and only perform the yield if this elapsed time gets close to the seastar reactor stall warning threshold.

Reply:
Or `ss::do_for_each`

Reply:
The maybe_yield function won't actually yield back to seastar after every iteration. It first checks to see how long the fiber has been running first and only yields once the fiber is close to or exceeding it's recommended time slice. So I don't think there is any need to duplicate that logic in the function.

Reply:
Ah now that i am re-reading this I assumed incorrectly. Nice !

Reply:
It's possible. Right now the issue is de-allocating the `_logs` container after all of the logs had `closed()` called on them. What would be gained by closing them in batches?

Reply:
I thought `close()` was being invoked on destruction. What is actually taking so long in the first place?